### PR TITLE
Containers: test privileged mode

### PR DESCRIPTION
**Pull request by ilausuch**
Test privileged mode in podman and docker. The command mount -t tmpfs none /mnt
only works in privileged mode because the read-only protection in the default mode

- Related ticket: https://progress.opensuse.org/issues/135518
- Verification run: 
- http://openqa.suse.de/tests/12335440 (15-SP4 podman)
- http://openqa.suse.de/tests/12335443 (15-SP4 docker)
- http://openqa.suse.de/tests/12335444 (12-SP5 docker)

### CODE REVIEW

File: tests/containers/privileged_mode.pm [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    my $engine = $self->containers_factory($runtime);
+    $self->{runtime} = $engine;
+    reset_container_network_if_needed($runtime);
+
+    my $image = get_var("CONTAINER_IMAGE_TO_TEST", "registry.suse.com/bci/bci-base:latest");
+
+    record_info('Test', 'Launch a container with privileged mode');
+    assert_script_run("$runtime run -ti --rm --privileged $image ls /dev");
+    assert_script_run("$runtime run -ti --rm --privileged $image ls /proc/1/ns/");
+    assert_script_run("$runtime run -ti --rm --privileged $image mount -t tmpfs none /mnt");
```
Comment (pdostal) [inline tests/containers/privileged_mode.pm:30]:
```suggestion
    # Mounting tmpfs only works in privileged mode because the read-only protection in the default mode
    assert_script_run("$runtime run -ti --rm --privileged $image mount -t tmpfs none /mnt");
```

File: tests/containers/privileged_mode.pm [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    assert_script_run("modprobe dummy && lsmod | grep dummy");
+    
+    my $capbnd = script_output("cat /proc/1/status | grep CapBnd");
+    validate_script_output($runtime . " run -ti --rm --privileged $image cat /proc/1/status | grep CapBnd", sub { m/$capbnd/ });
```
Comment (pdostal) [inline tests/containers/privileged_mode.pm:34]:
```suggestion
    validate_script_output("$runtime run -ti --rm --privileged $image cat /proc/1/status | grep CapBnd", sub { m/$capbnd/ });
```

**Review by pdostal — APPROVED** [review]

File: tests/containers/privileged_mode.pm [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    $self->{runtime} = $engine;
+    reset_container_network_if_needed($runtime);
+
+    my $image = get_var("CONTAINER_IMAGE_TO_TEST", "registry.suse.com/bci/bci-base:latest");
+
+    record_info('Test', 'Launch a container with privileged mode');
+    assert_script_run("$runtime run -ti --rm --privileged $image ls /dev");
+    assert_script_run("$runtime run -ti --rm --privileged $image ls /proc/1/ns/");
+    assert_script_run("$runtime run -ti --rm --privileged $image mount -t tmpfs none /mnt");
+    assert_script_run("modprobe dummy && lsmod | grep dummy");
```
Comment (mloviska) [inline tests/containers/privileged_mode.pm:31]: why is this needed?

File: tests/containers/privileged_mode.pm (hunk 1 above)
Comment (ilausuch) [reply tests/containers/privileged_mode.pm:30]: do you mean to explain all cases? ok

File: tests/containers/privileged_mode.pm (hunk 3 above)
Comment (ilausuch) [reply tests/containers/privileged_mode.pm:31]: I wanted to add all the possibilities, maybe this is not interesting. I can remove this, it's not a problem. Only I will wait a little to others thoughts

File: tests/containers/privileged_mode.pm [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+    my $runtime = $args->{runtime};
+    my $engine = $self->containers_factory($runtime);
+    $self->{runtime} = $engine;
+    reset_container_network_if_needed($runtime);
+
+    my $image = get_var("CONTAINER_IMAGE_TO_TEST", "registry.suse.com/bci/bci-base:latest");
+
+    record_info('Test', 'Launch a container with privileged mode');
+    assert_script_run("$runtime run -ti --rm --privileged $image ls /dev");
```
Comment (ricardobranco777) [inline tests/containers/privileged_mode.pm:28]: Unprivileged containers also have a /dev directory.  I think this line should check for devices that typically appear in the host's /dev but not on unprivileged's like /dev/bus:

```suggestion
    assert_script_run("[ $($runtime run -ti --rm --privileged $image ls /dev/bus");
```

File: tests/containers/privileged_mode.pm (hunk 3 above)
Comment (ricardobranco777) [reply tests/containers/privileged_mode.pm:31]: This pipeline should be run inside the container.

File: tests/containers/privileged_mode.pm [hunk 5]
```diff
@@ -0,0 +1,52 @@
+# SUSE's openQA tests
+#
+# Copyright 2023 SUSE LLC
+# SPDX-License-Identifier: FSFAP
+
+# Package: podman
+# Summary: Test podman with systemd
```
Comment (ricardobranco777) [inline tests/containers/privileged_mode.pm:7]: Let's change the summary.

```suggestion
# Summary: Check your privilege
```

File: tests/containers/privileged_mode.pm [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+sub run {
+    my ($self, $args) = @_;
+    select_serial_terminal;
+
+    my $runtime = $args->{runtime};
+    my $engine = $self->containers_factory($runtime);
+    $self->{runtime} = $engine;
+    reset_container_network_if_needed($runtime);
+
+    my $image = get_var("CONTAINER_IMAGE_TO_TEST", "registry.suse.com/bci/bci-base:latest");
```
Comment (grisu48) [inline tests/containers/privileged_mode.pm:25]: This is a test of the container engine. Those test runs don't use `CONTAINER_IMAGE_TO_TEST` because they are testing the engine and not an image.

Please remove the `get_var` and just use a static image, or use a different variable.

File: tests/containers/privileged_mode.pm [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+    record_info('Test', 'Launch a container with privileged mode');
+    assert_script_run("$runtime run -ti --rm --privileged $image ls /dev");
+    assert_script_run("$runtime run -ti --rm --privileged $image ls /proc/1/ns/");
```
Comment (grisu48) [inline tests/containers/privileged_mode.pm:29]: This also works for unprivileged containers.

File: tests/containers/privileged_mode.pm (hunk 3 above)
Comment (grisu48) [reply tests/containers/privileged_mode.pm:31]: Also: `modprobe` is not available in `registry.suse.com/bci/bci-base`.

Comment (ilausuch) [reply tests/containers/privileged_mode.pm:31]: ahhh yes, I understand, @ricardobranco777 pointed that! this has to be inside the container to make any sence

**Comment by ilausuch** [comment]
Thank you all for your reviews

File: tests/containers/privileged_mode.pm (hunk 4 above)
Comment (ilausuch) [reply tests/containers/privileged_mode.pm]: done

File: tests/containers/privileged_mode.pm (hunk 6 above)
Comment (ilausuch) [reply tests/containers/privileged_mode.pm]: done

File: tests/containers/privileged_mode.pm (hunk 7 above)
Comment (ilausuch) [reply tests/containers/privileged_mode.pm]: removed

**Comment by ilausuch** [comment]
All new VR ready

File: tests/containers/privileged_mode.pm [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    my $runtime = $args->{runtime};
+    my $engine = $self->containers_factory($runtime);
+    $self->{runtime} = $engine;
+    reset_container_network_if_needed($runtime);
+
+    my $image = "registry.suse.com/bci/bci-base:latest";
+
+    record_info('Test', 'Launch a container with privileged mode');
+    # /dev is only accessible in privileged mode
+    assert_script_run("$runtime run -ti --rm --privileged $image ls /dev/bus");
```
Comment (ricardobranco777) [inline tests/containers/privileged_mode.pm:29]:
```suggestion
    assert_script_run("$runtime run --rm --privileged $image ls /dev/bus");
```

File: tests/containers/privileged_mode.pm [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+    # Mounting tmpfs only works in privileged mode because the read-only protection in the default mode
+    assert_script_run("$runtime run -ti --rm --privileged $image mount -t tmpfs none /mnt");
```
Comment (ricardobranco777) [inline tests/containers/privileged_mode.pm:32]:
```suggestion
    assert_script_run("$runtime run --rm --privileged $image mount -t tmpfs none /mnt");
```

File: tests/containers/privileged_mode.pm [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+    # Capabilities are only available in privileged mode
+    my $capbnd = script_output("cat /proc/1/status | grep CapBnd");
+    validate_script_output("$runtime run -ti --rm --privileged $image cat /proc/1/status | grep CapBnd", sub { m/$capbnd/ });
```
Comment (ricardobranco777) [inline tests/containers/privileged_mode.pm:36]:
```suggestion
    validate_script_output("$runtime run --rm --privileged $image cat /proc/1/status | grep CapBnd", sub { m/$capbnd/ });
```

**Review by ricardobranco777 — APPROVED** [review]
Please drop `-ti`.  Otherwise LGTM.

**Comment by ilausuch** [comment]
Thanks @ricardobranco777 changes done. Last VR for this change http://openqa.suse.de/tests/12336305

**Comment by pdostal** [comment]
Yesterday on s390x (see [example](https://openqa.suse.de/tests/12331387#step/docker_privileged_mode/68) and [Slack thread](https://suse.slack.com/archives/C02CGKBCGT1/p1695965543403489)):
```
ls: cannot access '/dev/bus': No such file or directory
```
